test/Microsoft.AspNetCore.Server.Kestrel.FunctionalTests/TestServer.cs

```diff
@@ -47,63 +49,68 @@ public TestServer(RequestDelegate app, TestServiceContext context, ListenOptions
 
         public TestServer(RequestDelegate app, TestServiceContext context, ListenOptions listenOptions, IHttpContextFactory httpContextFactory)
         {
+            _app = app;
             _listenOptions = listenOptions;
-
             Context = context;
-            context.ServerOptions.ListenOptions.Add(_listenOptions);
-
-            // Switch this to test on socket transport
-            var transportFactory = CreateLibuvTransportFactory(context);
-            // var transportFactory = CreateSocketTransportFactory(context);
-
-            _server = new KestrelServer(transportFactory, context);
-            var httpApplication = new DummyApplication(app, httpContextFactory);
-
-            try
-            {
-                _server.StartAsync(httpApplication, CancellationToken.None).Wait();
-            }
-            catch
-            {
-                _server.StopAsync(new CancellationTokenSource(_shutdownTimeout).Token).Wait();
-                _server.Dispose();
-                throw;
-            }
-        }
 
-        private static ITransportFactory CreateLibuvTransportFactory(TestServiceContext context)
-        {
-            var transportOptions = new LibuvTransportOptions { ThreadCount = 1 };
-
-            var transportFactory = new LibuvTransportFactory(
-                Options.Create(transportOptions),
-                new LifetimeNotImplemented(),
-                new KestrelTestLoggerFactory(new TestApplicationErrorLogger()));
-
-            return transportFactory;
+            _host = new WebHostBuilder()
+                     .UseKestrel(o =>
+                     {
+                         o.ListenOptions.Add(_listenOptions);
+                     })
+                     .ConfigureServices(services =>
+                     {
+                         if (httpContextFactory != null)
+                         {
+                             services.AddSingleton(httpContextFactory);
+                         }
+
+                         services.AddSingleton<IStartup>(this);
+                         services.AddSingleton<ILoggerFactory>(new KestrelTestLoggerFactory(context.ErrorLogger));
+                         services.AddSingleton<IServer>(sp =>
+                         {
+                             // Manually configure options on the TestServiceContext.
+                             // We're doing this so we can use the same instance that was passed in
+                             var configureOptions = sp.GetServices<IConfigureOptions<KestrelServerOptions>>();
+                             foreach (var c in configureOptions)
+                             {
+                                 c.Configure(context.ServerOptions);
```

Comment:
`o.ListenOptions.Add(_listenOptions);` is the only thing configured, right? Why not just call `context.ServerOptions.ListenOptions.Add(_listenOptions);` here?